File: command/java/Command.java
// 命令执行者
class Button {
    private final Command command;

    public Button(Command command) {
        this.command = command;
    }

    public void press() {
        command.execute();
    }
}

// 命令
public interface Command {
    void execute();
}

class TurnOnCommand implements Command {
    private final Device device;

    public TurnOnCommand(Device device) {
        this.device = device;
    }

    @Override
    public void execute() {
        device.turnOn();
    }
}

class TurnOffCommand implements Command {
    private final Device device;

    public TurnOffCommand(Device device) {
        this.device = device;
    }

    @Override
    public void execute() {
        device.turnOff();
    }
}

class VolumeUpCommand implements Command {
    private final Device device;

    public VolumeUpCommand(Device device) {
        this.device = device;
    }

    @Override
    public void execute() {
        device.volumeUp();
    }
}

class VolumeDownCommand implements Command {
    private final Device device;

    public VolumeDownCommand(Device device) {
        this.device = device;
    }

    @Override
    public void execute() {
        device.volumeDown();
    }
}

interface Device {
    void turnOn(); // 开机

    void turnOff(); // 关机

    void volumeUp(); // 音量 +

    void volumeDown(); // 音量 -
}

class TV implements Device {
    private boolean isRunning;
    private int volume;

    @Override
    public void turnOn() {
        if (!isRunning) {
            isRunning = true;
        }
        System.out.println("Turning tv on");
    }

    @Override
    public void turnOff() {
        if (isRunning) {
            isRunning = false;
        }
        System.out.println("Turning tv off");
    }

    @Override
    public void volumeUp() {
        if (volume < 30) {
            ++volume;
        }
        System.out.println("Volume up to " + volume);
    }

    @Override
    public void volumeDown() {
        if (volume > 0) {
            --volume;
        }
        System.out.println("Volume down to " + volume);
    }
}

File: command/go/command.go
package command

import "fmt"

// Button 命令执行者
type Button struct {
	command Command
}

func NewButton(command Command) *Button {
	return &Button{command}
}

func (b *Button) Press() {
	b.command.Execute()
}

// Command 命令
type Command interface {
	Execute()
}

// 设备命令（具体命令）
type deviceCommand func()

func (cmd *deviceCommand) Execute() {
	(*cmd)()
}

type (
	TurnOnCommand     = deviceCommand // 设备开机命令
	TurnOffCommand    = deviceCommand // 设备关机命令
	VolumeUpCommand   = deviceCommand // 设备音量+命令
	VolumeDownCommand = deviceCommand // 设备音量-命令
)

// Device 命令接收接口，抽象「设备」
type Device interface {
	TurnOn()     // 开机
	TurnOff()    // 关机
	VolumeUp()   // 音量 +
	VolumeDown() // 音量 -
}

// TV 具体「设备」：电视机
type TV struct {
	isRunning bool
	volume    uint8
}

func (tv *TV) TurnOn() {
	if !tv.isRunning {
		tv.isRunning = true
	}
	fmt.Println("Turning tv on")
}

func (tv *TV) TurnOff() {
	if tv.isRunning {
		tv.isRunning = false
	}
	fmt.Println("Turning tv off")
}

func (tv *TV) VolumeUp() {
	if tv.volume < 30 {
		tv.volume += 1
	}
	fmt.Println("Volume up to", tv.volume)
}

func (tv *TV) VolumeDown() {
	if tv.volume > 0 {
		tv.volume -= 1
	}
	fmt.Println("Volume down to", tv.volume)
}
